README
NAME
    "Socket::Packet" - interface to Linux's "PF_PACKET" socket family

SYNOPSIS
     use Socket qw( SOCK_RAW );
     use Socket::Packet qw(
        PF_PACKET
        ETH_P_ALL
        pack_sockaddr_ll unpack_sockaddr_ll
     );
 
     socket( my $sock, PF_PACKET, SOCK_RAW, 0 )
        or die "Cannot socket() - $!\n";
 
     bind( $sock, pack_sockaddr_ll( ETH_P_ALL, 0, 0, 0, "" ) )
        or die "Cannot bind() - $!\n";
 
     while( my $addr = recv( $sock, my $packet, 8192, 0 ) ) {
        my ( $proto, $ifindex, $hatype, $pkttype, $addr )
           = unpack_sockaddr_ll( $addr );

        ...
     }

DESCRIPTION
    To quote packet(7):

     Packet sockets are used to receive or send raw packets at the device driver
     (OSI Layer 2) level. They allow the user to implement protocol modules in
     user space on top of the physical layer.

    Sockets in the "PF_PACKET" family get direct link-level access to the
    underlying hardware (i.e. Ethernet or similar). They are usually used to
    implement packet capturing, or sending of specially-constructed packets
    or to implement protocols the underlying kernel does not recognise.

    The use of "PF_PACKET" sockets is usually restricted to privileged users
    only.

    This module also provides various other support functions which wrap
    "ioctl()"s or socket options. This includes support for
    "PACKET_RX_RING", the high-performance zero-copy packet receive
    buffering, if the underlying platform supports it.

CONSTANTS
    The following constants are exported

    PF_PACKET
            The packet family (for "socket()" calls)

    AF_PACKET
            The address family

    PACKET_HOST
            This packet is inbound unicast for this host.

    PACKET_BROADCAST
            This packet is inbound broadcast.

    PACKET_MULTICAST
            This packet is inbound multicast.

    PACKET_OTHERHOST
            This packet is inbound unicast for another host.

    PACKET_OUTGOING
            This packet is outbound.

    ETH_P_ALL
            Pseudo-protocol number to capture all protocols.

    SOL_PACKET
            Socket option level for "getsockopt" and "setsockopt".

SOCKET OPTIONS
    The following constants define socket options

    PACKET_STATISTICS (get; struct tpacket_stats)
            Packet received and drop counters.

    PACKET_ORIGDEV (get or set; int)
            Received packets will indicate the originally-received device,
            rather than the apparent one. This mainly relates to Ethernet
            bonding or VLANs.

            This socket option is optional, and may not be provided on all
            platforms.

    PACKET_ADD_MEMBERSHIP (set; struct packet_mreq)
    PACKET_DROP_MEMBERSHIP (set; struct packet_mreq)
            Membership of multicast or broadcast groups, or set promiscuous
            mode.

            The "packet_mreq" "type" field should be one of the following:

            PACKET_MR_MULTICAST
                A multicast group

            PACKET_MR_PROMISC
                Set or clear the promiscuous flag; the address is ignored

            PACKET_MR_ALLMULTI
                Set or clear the allmulti flag; the address is ignored

FUNCTIONS
    The following pair of functions operate on "AF_PACKET" address
    structures. The meanings of the parameters are:

    protocol
            An ethertype protocol number. When using an address with
            "bind()", the constant "ETH_P_ALL" can be used instead, to
            capture any protocol. The "pack_sockaddr_ll()" and
            "unpack_sockaddr_ll()" functions byte-swap this value to or from
            network endian order.

    ifindex The index number of the interface on which the packet was sent
            or received. When using an address with "bind()", the value 0
            can be used instead, to watch all interfaces.

    hatype  The hardware ARP type of hardware address.

    pkttype The type of the packet; indicates if it was sent or received.
            Will be one of the "PACKET_*" values.

    addr    The underlying hardware address, in the type given by "hatype".

  $a = pack_sockaddr_ll( $protocol, $ifindex, $hatype, $pkttype, $addr )
    Returns a "sockaddr_ll" structure with the fields packed into it.

  ( $protocol, $ifindex, $hatype, $pkttype, $addr ) = unpack_sockaddr_ll( $a )
    Takes a "sockaddr_ll" structure and returns the unpacked fields from it.

  $mreq = pack_packet_mreq( $ifindex, $type, $addr )
    Returns a "packet_mreq" structure with the fields packed into it.

  ( $ifindex, $type, $addr ) = unpack_packet_mreq( $mreq )
    Takes a "packet_mreq" structure and returns the unpacked fields from it.

  ( $packets, $drops ) = unpack_tpacket_stats( $stats )
    Takes a "tpacket_stats" structure from the "PACKET_STATISTICS" sockopt
    and returns the unpacked fields from it.

  $time = siocgstamp( $sock )
  ( $sec, $usec ) = siocgstamp( $sock )
    Returns the timestamp of the last received packet on the socket (as
    obtained by the "SIOCGSTAMP" "ioctl"). In scalar context, returns a
    single floating-point value in UNIX epoch seconds. In list context,
    returns the number of seconds, and the number of microseconds.

  $time = siocgstampns( $sock )
  ( $sec, $nsec ) = siocgstampns( $sock )
    Returns the nanosecond-precise timestamp of the last received packet on
    the socket (as obtained by the "SIOCGSTAMPNS" "ioctl"). In scalar
    context, returns a single floating-point value in UNIX epoch seconds. In
    list context, returns the number of seconds, and the number of
    nanoseconds.

  $ifindex = siocgifindex( $sock, $ifname )
    Returns the "ifindex" of the interface with the given name if one
    exists, or "undef" if not. $sock does not need to be a "PF_PACKET"
    socket, any socket handle will do.

  $ifname = siocgifname( $sock, $ifindex )
    Returns the "ifname" of the interface at the given index if one exists,
    or "undef" if not. $sock does not need to be a "PF_PACKET" socket, any
    socket handle will do.

  ( $addr, $len ) = recv_len( $sock, $buffer, $maxlen, $flags )
    Similar to Perl's "recv" builtin, except it returns the packet length as
    an explict return value. This may be useful if $flags contains the
    "MSG_TRUNC" flag, obtaining the true length of the packet on the wire,
    even if this is longer than the data written in the buffer.

RING-BUFFER FUNCTIONS
    The following functions operate on the high-performance memory-mapped
    buffer feature of "PF_PACKET", allowing efficient packet-capture
    applications to share a buffer with the kernel directly, avoiding the
    need for per-packet system calls to "recv()" (and possibly "ioctl()" to
    obtain the timestamp).

    The ring-buffer is optional, and may not be implemented on all
    platforms. If it is not implemented, then all the following functions
    will die with an error message.

  $size = setup_rx_ring( $sock, $frame_size, $frame_nr, $block_size )
    Sets up the ring-buffer on the socket. The buffer will store $frame_nr
    frames of up to $frame_size bytes each (including metadata headers), and
    will be split in the kernel in blocks of $block_size bytes. $block_size
    should be a power of 2, at minimum, 4KiB.

    If successful, the overall size of the buffer in bytes is returned. If
    not, "undef" is returned, and $! will hold the error value.

  $status = get_ring_frame_status( $sock )
    Returns the frame status of the next frame in the ring.

    The following constants are defined for the status:

    TP_STATUS_KERNEL
            This frame belongs to the kernel and userland should not touch
            it.

    TP_STATUS_USER
            This frame belongs to userland and the kernel will not modify
            it.

    TP_STATUS_LOSING
            Bitwise-or'ed with the status if packet loss has occurred since
            the previous frame.

  $len = get_ring_frame( $sock, $buffer, \%info )
    If the next frame is ready for userland, fills in keys of the %info hash
    with its metadata, sets $buffer to its contents, and return the length
    of the data. The $buffer variable has its string backing buffer aliased,
    rather than the buffer copied into, for performance. The caller should
    not modify the variable, nor attempt to access it after the socket has
    been closed.

    If the frame is not yet ready, this function returns undef.

    The following fields are returned:

    tp_status
            The status of the frame; see "get_ring_frame_status()"

    tp_len  The length of the packet on the wire, in bytes

    tp_snaplen
            The length of the packet captured and stored in the buffer, in
            bytes. This may be shorter than "tp_len" if, for example, a
            filter is set on the socket that truncated the packet.

    tp_sec
    tp_nsec The seconds and nanoseconds fields of the timestamp. If the
            underlying platform does not support "TPACKET_V2", then this
            field will only have a resolution of microseconds; i.e. it will
            be a whole multiple of 1000.

    tp_vlan_tci
            VLAN information about the packet, if the underlying platform
            supports "TPACKET_V2". If this is not supported, the key will
            not be present in the hash

    sll_protocol
    sll_ifindex
    sll_hatype
    sll_pkttype
    sll_addr
            Fields from the "struct sockaddr_ll"; see above for more detail

  clear_ring_frame( $sock )
    Clears the status of current frame to hand it back to the kernel and
    moves on to the next.

SEE ALSO
    *   IO::Socket::Packet - Object interface to "AF_PACKET" domain sockets

    *   Linux::SocketFilter - interface to Linux's socket packet filtering

    *   packet(7) - packet, AF_PACKET - packet interface on device level

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Socket-Packet

0.09    CHANGES:
         * Better Linux version test in t/13recv_len.t

0.08    CHANGES:
         * Use new ability of ExtUtils::CChecker 0.04 to look harder to find
           header files suitable for TPACKET_V2

0.07    CHANGES:
         * Support socket options:
            + PACKET_STATISTICS
            + PACKET_ORIGDEV
            + PACKET_ADD_MEMBERSHIP and PACKET_DROP_MEMBERSHIP
            + PACKET_RX_RING and related
         * Provide proper Perl-level access to the RX ring buffer

0.06    CHANGES:
         * Implement ->recv_unpack using ->recv_len so it yields wire packet
           length and doesn't crash with MSG_TRUNC flag

        BUGFIXES:
         * Fix recv_len() so it works on Perl 5.8

0.05    CHANGES:
         * Added recv_len() function and method

0.04    CHANGES:
         * Use XSLoader instead of DynaLoader
         * import Exporter::import instead of @ISAing it
         * Use ExtUtils::CChecker for build checking

        BUGFIXES:
         * Run fail tests against a pipe to hopefully avoid spurious packet-
           related side-effects on INET sockets

0.03    CHANGES:
         * Wrap SIOCGIFINDEX and SIOCGIFNAME to provide name <=> index lookup
         * Provided an ARP example to demonstrate packet sending
         * Better 'OS unsupported' line if ExtUtils::CBuilder is missing

0.02    CHANGES:
         * Wrap SIOCGSTAMP and SIOCGSTAMPNS to obtain packet receive timestamp
         * Created IO::Socket wrapping under the name IO::Socket::Packet

        BUGFIXES:
         * More robust Build.PL, includes OS capability detection

0.01    First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;

use Module::Build;

eval { require ExtUtils::CChecker; 1 } or
   die "OS unsupported - missing ExtUtils::CChecker";
eval { ExtUtils::CChecker->VERSION( 0.04 ) } or
   die "OS unsupported - ExtUtils::CChecker too old; need at least 0.04";

my $cc = ExtUtils::CChecker->new;

$cc->assert_compile_run(
   diag => "no PF_PACKET",
   source => <<'EOF' );
#include <sys/socket.h>
int family = PF_PACKET;
int main(int argc, char *argv[]) {
  return 0;
}
EOF

# Some userland headers are sufficiently old that they don't have TPACKET_V2
# or the tpacket2_hdr structure. If this is the case we have little choice but
# to cheat a little and try to use the kernel ones instead.
chomp( my $uname_r = `uname -r` );

my @dirs = (
   [],
   [ "/lib/modules/$uname_r/source/include" ],
);

$cc->find_include_dirs_for(
   diag => "no PACKET_STATISTICS",
   dirs => \@dirs,
   source => <<'EOF' );
#include <linux/if_packet.h>
int origdev = PACKET_STATISTICS;
struct tpacket_stats stats;
int main(int argc, char *argv[]) {
  return 0;
}
EOF

$cc->try_find_include_dirs_for(
   define => "HAVE_ORIGDEV",
   dirs => \@dirs,
   source => <<'EOF' );
#include <linux/if_packet.h>
int origdev = PACKET_ORIGDEV;
int main(int argc, char *argv[]) {
  return 0;
}
EOF

$cc->try_find_include_dirs_for(
   define => "HAVE_TPACKET",
   dirs => \@dirs,
   source => <<'EOF' );
#include <linux/if_packet.h>
struct tpacket_hdr hdr;
int main(int argc, char *argv[]) {
  return 0;
}
EOF

$cc->try_find_include_dirs_for(
   define => "HAVE_TPACKET2",
   dirs => \@dirs,
   source => <<'EOF' );
#include <linux/if_packet.h>
int pkttype = TPACKET_V2;
struct tpacket2_hdr hdr;
int main(int argc, char *argv[]) {
  return 0;
}
EOF

my $build = $cc->new_module_build(
   module_name => 'Socket::Packet',
   requires => {
   },
   configure_requires => {
      'ExtUtils::CChecker' => '0.04',
      'Module::Build' => 0,
      'Module::Build::Compat' => 0,
   },
   build_requires => {
      'ExtUtils::CChecker' => 0,
      'IO::Socket::INET' => 0,
      'Module::Build' => 0,
      'Module::Build::Compat' => 0,
      'Test::More' => 0,
   },
   license => 'perl',
   create_makefile_pl => 'small',
   create_license => 1,
   create_readme  => 1,
);

$build->create_build_script;


examples/capture.pl
#!/usr/bin/perl

use strict;
use warnings;

use IO::Socket::Packet;
use Socket::Packet qw(
   PACKET_OUTGOING
);

my $sock = IO::Socket::Packet->new( IfIndex => 0 )
   or die "Cannot create PF_PACKET socket - $!";

while( my ( $proto, $ifindex, $hatype, $pkttype, $addr ) = $sock->recv_unpack( my $packet, 8192, 0 ) ) {
   my ( $ts, $ts_usec ) = $sock->timestamp;
   my @ts = localtime $ts;
   printf "[%4d/%02d/%02d %02d:%02d:%02d.%06d] ", $ts[5]+1900, @ts[4,3,2,1,0], $ts_usec;

   # Reformat nicely for printing
   $addr = join( ":", map sprintf("%02x", ord $_), split //, $addr );

   if( $pkttype == PACKET_OUTGOING ) {
      print "Sent a packet to $addr";
   }
   else {
      print "Received a packet from $addr";
   }

   printf " of protocol %04x on %s:\n", $proto, $sock->ifindex2name( $ifindex );

   printf "  %v02x\n", $1 while $packet =~ m/(.{1,16})/sg;
}


examples/capture-rxring.pl
#!/usr/bin/perl

use strict;
use warnings;

use IO::Socket::Packet;
use Socket::Packet qw(
   PACKET_OUTGOING
);

my $sock = IO::Socket::Packet->new( IfIndex => 0 )
   or die "Cannot create PF_PACKET socket - $!";

unless( eval { $sock->setup_rx_ring( 2048, 128, 16384 ) } ) {
   die "Cannot setup PACKET_RX_RING - $@" if $@;
   die "Cannot setup PACKET_RX_RING - $!";
}

while( defined $sock->wait_ring_frame( my $packet, \my %info ) ) {
   my @ts = localtime $info{tp_sec};
   printf "[%4d/%02d/%02d %02d:%02d:%02d.%09d] ", $ts[5]+1900, @ts[4,3,2,1,0], $info{tp_nsec};

   # Reformat nicely for printing
   my $addr = join( ":", map sprintf("%02x", ord $_), split //, $info{sll_addr} );

   if( $info{sll_pkttype} == PACKET_OUTGOING ) {
      print "Sent a packet to $addr";
   }
   else {
      print "Received a packet from $addr";
   }

   printf " of protocol %04x on %s:\n", $info{sll_protocol}, $sock->ifindex2name( $info{sll_ifindex} );

   printf "  %v02x\n", $1 while $packet =~ m/(.{1,16})/sg;

   $sock->done_ring_frame;
}


lib/Socket/Packet.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2009-2012 -- leonerd@leonerd.org.uk

package Socket::Packet;

use strict;
use warnings;

use Carp;

our $VERSION = '0.09';

use Exporter 'import';
our @EXPORT_OK = qw(
   pack_sockaddr_ll unpack_sockaddr_ll
   pack_packet_mreq unpack_packet_mreq
   unpack_tpacket_stats
   siocgstamp
   siocgstampns
   siocgifindex
   siocgifname
   recv_len

   setup_rx_ring
   get_ring_frame_status
   get_ring_frame
   done_ring_frame
);

require XSLoader;
XSLoader::load( __PACKAGE__, $VERSION );

=head1 NAME

C<Socket::Packet> - interface to Linux's C<PF_PACKET> socket family

=head1 SYNOPSIS

 use Socket qw( SOCK_RAW );
 use Socket::Packet qw(
    PF_PACKET
    ETH_P_ALL
    pack_sockaddr_ll unpack_sockaddr_ll
 );
 
 socket( my $sock, PF_PACKET, SOCK_RAW, 0 )
    or die "Cannot socket() - $!\n";
 
 bind( $sock, pack_sockaddr_ll( ETH_P_ALL, 0, 0, 0, "" ) )
    or die "Cannot bind() - $!\n";
 
 while( my $addr = recv( $sock, my $packet, 8192, 0 ) ) {
    my ( $proto, $ifindex, $hatype, $pkttype, $addr )
       = unpack_sockaddr_ll( $addr );

    ...
 }

=head1 DESCRIPTION

To quote C<packet(7)>:

 Packet sockets are used to receive or send raw packets at the device driver
 (OSI Layer 2) level. They allow the user to implement protocol modules in
 user space on top of the physical layer.

Sockets in the C<PF_PACKET> family get direct link-level access to the
underlying hardware (i.e. Ethernet or similar). They are usually used to
implement packet capturing, or sending of specially-constructed packets
or to implement protocols the underlying kernel does not recognise.

The use of C<PF_PACKET> sockets is usually restricted to privileged users
only.

This module also provides various other support functions which wrap
C<ioctl()>s or socket options. This includes support for C<PACKET_RX_RING>,
the high-performance zero-copy packet receive buffering, if the underlying
platform supports it.

=cut

=head1 CONSTANTS

The following constants are exported

=over 8

=item PF_PACKET

The packet family (for C<socket()> calls)

=item AF_PACKET

The address family

=item PACKET_HOST

This packet is inbound unicast for this host.

=item PACKET_BROADCAST

This packet is inbound broadcast.

=item PACKET_MULTICAST

This packet is inbound multicast.

=item PACKET_OTHERHOST

This packet is inbound unicast for another host.

=item PACKET_OUTGOING

This packet is outbound.

=item ETH_P_ALL

Pseudo-protocol number to capture all protocols.

=item SOL_PACKET

Socket option level for C<getsockopt> and C<setsockopt>.

=back

=cut

=head1 SOCKET OPTIONS

The following constants define socket options

=over 8

=item PACKET_STATISTICS (get; struct tpacket_stats)

Packet received and drop counters.

=item PACKET_ORIGDEV (get or set; int)

Received packets will indicate the originally-received device, rather than the
apparent one. This mainly relates to Ethernet bonding or VLANs.

This socket option is optional, and may not be provided on all platforms.

=item PACKET_ADD_MEMBERSHIP (set; struct packet_mreq)

=item PACKET_DROP_MEMBERSHIP (set; struct packet_mreq)

Membership of multicast or broadcast groups, or set promiscuous mode.

The C<packet_mreq> C<type> field should be one of the following:

=over 4

=item PACKET_MR_MULTICAST

A multicast group

=item PACKET_MR_PROMISC

Set or clear the promiscuous flag; the address is ignored

=item PACKET_MR_ALLMULTI

Set or clear the allmulti flag; the address is ignored

=back

=back

=cut

=head1 FUNCTIONS

The following pair of functions operate on C<AF_PACKET> address structures.
The meanings of the parameters are:

=over 8

=item protocol

An ethertype protocol number. When using an address with C<bind()>, the
constant C<ETH_P_ALL> can be used instead, to capture any protocol. The
C<pack_sockaddr_ll()> and C<unpack_sockaddr_ll()> functions byte-swap this
value to or from network endian order.

=item ifindex

The index number of the interface on which the packet was sent or received.
When using an address with C<bind()>, the value C<0> can be used instead, to
watch all interfaces.

=item hatype

The hardware ARP type of hardware address.

=item pkttype

The type of the packet; indicates if it was sent or received. Will be one of
the C<PACKET_*> values.

=item addr

The underlying hardware address, in the type given by C<hatype>.

=back

=head2 $a = pack_sockaddr_ll( $protocol, $ifindex, $hatype, $pkttype, $addr )

Returns a C<sockaddr_ll> structure with the fields packed into it.

=head2 ( $protocol, $ifindex, $hatype, $pkttype, $addr ) = unpack_sockaddr_ll( $a )

Takes a C<sockaddr_ll> structure and returns the unpacked fields from it.

=head2 $mreq = pack_packet_mreq( $ifindex, $type, $addr )

Returns a C<packet_mreq> structure with the fields packed into it.

=head2 ( $ifindex, $type, $addr ) = unpack_packet_mreq( $mreq )

Takes a C<packet_mreq> structure and returns the unpacked fields from it.

=head2 ( $packets, $drops ) = unpack_tpacket_stats( $stats )

Takes a C<tpacket_stats> structure from the C<PACKET_STATISTICS> sockopt and
returns the unpacked fields from it.

=head2 $time = siocgstamp( $sock )

=head2 ( $sec, $usec ) = siocgstamp( $sock )

Returns the timestamp of the last received packet on the socket (as obtained
by the C<SIOCGSTAMP> C<ioctl>). In scalar context, returns a single
floating-point value in UNIX epoch seconds. In list context, returns the
number of seconds, and the number of microseconds.

=head2 $time = siocgstampns( $sock )

=head2 ( $sec, $nsec ) = siocgstampns( $sock )

Returns the nanosecond-precise timestamp of the last received packet on the
socket (as obtained by the C<SIOCGSTAMPNS> C<ioctl>). In scalar context,
returns a single floating-point value in UNIX epoch seconds. In list context,
returns the number of seconds, and the number of nanoseconds.

=head2 $ifindex = siocgifindex( $sock, $ifname )

Returns the C<ifindex> of the interface with the given name if one exists, or
C<undef> if not. C<$sock> does not need to be a C<PF_PACKET> socket, any
socket handle will do.

=head2 $ifname = siocgifname( $sock, $ifindex )

Returns the C<ifname> of the interface at the given index if one exists, or
C<undef> if not. C<$sock> does not need to be a C<PF_PACKET> socket, any
socket handle will do.

=head2 ( $addr, $len ) = recv_len( $sock, $buffer, $maxlen, $flags )

Similar to Perl's C<recv> builtin, except it returns the packet length as an
explict return value. This may be useful if C<$flags> contains the
C<MSG_TRUNC> flag, obtaining the true length of the packet on the wire, even
if this is longer than the data written in the buffer.

=cut

=head1 RING-BUFFER FUNCTIONS

The following functions operate on the high-performance memory-mapped buffer
feature of C<PF_PACKET>, allowing efficient packet-capture applications to
share a buffer with the kernel directly, avoiding the need for per-packet
system calls to C<recv()> (and possibly C<ioctl()> to obtain the timestamp).

The ring-buffer is optional, and may not be implemented on all platforms. If
it is not implemented, then all the following functions will die with an error
message.

=cut

=head2 $size = setup_rx_ring( $sock, $frame_size, $frame_nr, $block_size )

Sets up the ring-buffer on the socket. The buffer will store C<$frame_nr>
frames of up to C<$frame_size> bytes each (including metadata headers), and
will be split in the kernel in blocks of C<$block_size> bytes.  C<$block_size>
should be a power of 2, at minimum, 4KiB.

If successful, the overall size of the buffer in bytes is returned. If not,
C<undef> is returned, and C<$!> will hold the error value.

=head2 $status = get_ring_frame_status( $sock )

Returns the frame status of the next frame in the ring.

The following constants are defined for the status:

=over 8

=item TP_STATUS_KERNEL

This frame belongs to the kernel and userland should not touch it.

=item TP_STATUS_USER

This frame belongs to userland and the kernel will not modify it.

=item TP_STATUS_LOSING

Bitwise-or'ed with the status if packet loss has occurred since the previous
frame.

=back

=head2 $len = get_ring_frame( $sock, $buffer, \%info )

If the next frame is ready for userland, fills in keys of the C<%info> hash
with its metadata, sets C<$buffer> to its contents, and return the length of
the data. The C<$buffer> variable has its string backing buffer aliased,
rather than the buffer copied into, for performance. The caller should not
modify the variable, nor attempt to access it after the socket has been
closed.

If the frame is not yet ready, this function returns undef.

The following fields are returned:

=over 8

=item tp_status

The status of the frame; see C<get_ring_frame_status()>

=item tp_len

The length of the packet on the wire, in bytes

=item tp_snaplen

The length of the packet captured and stored in the buffer, in bytes. This may
be shorter than C<tp_len> if, for example, a filter is set on the socket that
truncated the packet.

=item tp_sec

=item tp_nsec

The seconds and nanoseconds fields of the timestamp. If the underlying
platform does not support C<TPACKET_V2>, then this field will only have a
resolution of microseconds; i.e. it will be a whole multiple of 1000.

=item tp_vlan_tci

VLAN information about the packet, if the underlying platform supports
C<TPACKET_V2>. If this is not supported, the key will not be present in the
hash

=item sll_protocol

=item sll_ifindex

=item sll_hatype

=item sll_pkttype

=item sll_addr

Fields from the C<struct sockaddr_ll>; see above for more detail

=back

=head2 clear_ring_frame( $sock )

Clears the status of current frame to hand it back to the kernel and moves on
to the next.

=cut

=head1 SEE ALSO

=over 4

=item *

L<IO::Socket::Packet> - Object interface to C<AF_PACKET> domain sockets

=item *

L<Linux::SocketFilter> - interface to Linux's socket packet filtering

=item *

C<packet(7)> - packet, AF_PACKET - packet interface on device level

=back

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
